pyclam/manifold.py

```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
+        else:
+            raise ValueError(f'Cluster {cluster.name} is not in this edge.')
+
+
+class Graph:
+    """ An induced graph from clusters in the tree.
+
+    Clusters form the nodes, and two clusters with overlapping volumes have an edge connecting them.
+
+    The Graph Invariant:
+        * Every data point used to build the Manifold is also present in a Graph.
+        * Each point is present in exactly one Cluster.
+
+    This invariant only holds for full graphs, and not subgraph components.
+
     """
-    # TODO: Write dump/load methods for Graph.
+    # TODO: Implement dump/load
+
     def __init__(self, *clusters):
-        logging.debug(f'Graph(clusters={[str(c) for c in clusters]})')
-        assert all(isinstance(c, Cluster) for c in clusters)
-
-        # self.edges is a dictionary of:
-        #       the clusters in the graph, and
-        #       the set of edges from that cluster.
-        # An Edge is a named tuple of Neighbor, Distance, and Probability.
-        # Edge.neighbor is the neighboring cluster.
-        # Edge.distance is the distance to that neighbor.
-        # Edge.probability is used to pick an edge during a random walk.
-        self.edges: Dict[Cluster, Union[Set[Edge], None]] = {cluster: None for cluster in clusters}
+        logging.debug(f'Graph(clusters={list(map(str, clusters))})')
+        assert all(isinstance(c, Cluster) for c in clusters), 'all inputs to the Graph must be clusters.'
 
+        self.clusters: Set[Cluster] = {cluster for cluster in clusters}
+        self.edges: Set[Edge] = set()
+        self.is_built: bool = False  # this flag is set to True at the end of build_edges.
         self.cache: Dict[str, Any] = dict()
-        return
 
     def __eq__(self, other: 'Graph') -> bool:
-        """ Two graphs are identical if they have the same clusters and edges.
-        """
-        cluster_equality = set(self.clusters) == set(other.clusters)
-        if any((edges is None for edges in self.edges.values())):
-            return cluster_equality
-        else:
-            edges_equality = all((
-                set(self.edges[cluster]) == set(other.edges[cluster])
-                for cluster in self.clusters
-            ))
-            return cluster_equality and edges_equality
+        """ Two Graphs are identical if they have the same sets of clusters and edges. """
+        return (self.clusters == other.clusters) and (self.edges == other.edges)
 
     def __bool__(self) -> bool:
         return self.cardinality > 0
 
     def __iter__(self) -> Iterable[Cluster]:
-        """ An iterator over the clusters in the graph. """
-        yield from self.edges
+        yield from self.clusters
 
     def __str__(self) -> str:
-        # Cashing value because sort can be expensive on many clusters.
         if 'str' not in self.cache:
             self.cache['str'] = ', '.join(sorted(list(map(str, self.clusters))))
+        else:
+            pass
```

Comment:
why else pass? Seems overly verbose

Reply:
It's a Rust habit I subconsciously brought over. I didn't even notice it. Exhaustive if-branches are required in Rust and can be bad form in Python. I'm taking them all out of here